util/scfg/src/scfg.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "os/mynewt.h"
+#include "scfg/scfg.h"
+
+#define SCFG_MAX_SETTING_ID_LEN    64
+#define SCFG_MAX_NUM_STR_LEN       (sizeof "18446744073709551615")
+
+#define SCFG_FOREACH_SETTING(group, s) \
+    for (s = (group)->settings; s->name != NULL; s++)
+
+static void
+scfg_setting_id(const char *group_name, const char *setting_name, char *buf)
+{
+    int setting_len;
+    int group_len;
+    int off;
+
+    group_len = strlen(group_name);
+    setting_len = strlen(setting_name);
+
+    /* <group>/<setting> */
+    assert(group_len + 1 + setting_len <= SCFG_MAX_SETTING_ID_LEN);
+
+    off = 0;
+
+    memcpy(&buf[off], group_name, group_len);
+    off += group_len;
+
+    buf[off] = '/';
+    off++;
+
+    memcpy(&buf[off], setting_name, setting_len);
+    off += setting_len;
+
+    buf[off] = '\0';
+    off++;
+}
+
+static struct scfg_setting *
+scfg_find_setting_by_name(const struct scfg_group *group,
+                          const char *setting_name)
+{
+    const struct scfg_setting *setting;
+
+    SCFG_FOREACH_SETTING(group, setting) {
+        if (strcmp(setting->name, setting_name) == 0) {
+            return (struct scfg_setting *)setting;
```

Comment:
Does this cast need to be here?

Reply:
Yes, I believe it does.  The cast "casts away" const.  This is the same approach used by `strstr` for example,
```
char *strstr(const char *haystack, const char *needle);
```

Since C doesn't let you overload functions based on the "constness" of their parameters (not complaining!), we have to handle both const and non-const with the same function.  The approach that the standard library takes is to always accept const arguments and always return non-const.  It's not perfect, but I think it is reasonable.

Reply:
> The cast "casts away" const.

Ah, sorry, I totally missed the `const` keyword here, my bad!